### PR TITLE
Private courier 2

### PR DESCRIPTION
**Pull request by Sheodar**
Added courier for patreons.
1) Nobode can manage your courier.
2) If you have private courier, you can't manage team courier or other private courier.
2.1) All order received by the team's or alien courier will be transferred to the private courier.
2.2) You can't give item on team's or alien courier, if you have private courier (and if it alive)
3) Interface buttons "select courier" and "deliver things" rewrite. They work with the courier that is available to you.
4) Private couier has new model (lucky cat).

### CODE REVIEW

File: game/scripts/vscripts/addon_game_mode.lua [hunk 1]
```diff
@@ -649,6 +679,56 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
...
+	CustomGameEventManager:Send_ServerToPlayer(PlayerResource:GetPlayer(playerId), "selection_refresh", { pID = playerId })
+	if _G.personalCouriers[playerId] then
+		local privateCourier = _G.personalCouriers[playerId]
+
+		if orderType == DOTA_UNIT_ORDER_GIVE_ITEM and target:IsCourier() and target ~= privateCourier and privateCourier:IsAlive() then
+			CustomGameEventManager:Send_ServerToPlayer(PlayerResource:GetPlayer(playerId), "display_custom_error", { message = "#cannotgiveiteminthiscourier" })
+			return false
+		end
+
+		for i, x in pairs(filterTable.units) do
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
		for _, unitEntityIndex in pairs(filterTable.units) do
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 2]
```diff
@@ -649,6 +679,56 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
...
+			unit = EntIndexToHScript(filterTable.units[tostring(i)])
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
			local unit = EntIndexToHScript(unitEntityIndex)
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 3]
```diff
@@ -649,6 +679,56 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
...
+			if unit:IsCourier() and unit ~= _G.personalCouriers[playerId] and _G.personalCouriers[playerId]:IsAlive() then
+
+				local removeFocusEnt = { filterTable.units[i] }
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
				local removeFocusEnt = { unitEntityIndex }
```

File: content/panorama/layout/custom_game/common/utils.js [hunk 4]
```diff
@@ -66,3 +66,34 @@ function formatDate(date) {
...
+            if (p.id === 'Hud') {
+                return p;
+            } else {
+                p = p.GetParent();
+            }
+        }
+    } catch (e) {}
+}
+
+(function () {
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/utils.js]: It shouldn't be in utils

File: content/panorama/layout/custom_game/common/utils.js [hunk 5]
```diff
@@ -66,3 +66,34 @@ function formatDate(date) {
...
+    var playerId = Players.GetLocalPlayer()
+    var selectCourietButton = FindDotaHudElement('SelectCourierButton')
+    var deliverItemsButton = FindDotaHudElement('DeliverItemsButton')
+
+    selectCourietButton.SetPanelEvent("onactivate", function () {
+        GameEvents.SendCustomGameEventToServer("courier_custom_select", {playerId: playerId})
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/utils.js]: Don't send local player id as a custom payload, it's unnecessary and insecure

File: game/scripts/vscripts/addon_game_mode.lua [hunk 6]
```diff
@@ -649,6 +679,56 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
...
+					end
+				end
+
+				local newFocus = { privateCourier:GetEntityIndex() }
+				local haveCourierFocus = false
+				local selectionCounter = 0
+
+				Timers:CreateTimer(0.034, function()
+					if _G.selectionUnits[playerId] then
+						for i, x in pairs(_G.selectionUnits[playerId]) do
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
						for _, selectedEntityIndex in pairs(_G.selectionUnits[playerId]) do
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 7]
```diff
@@ -687,6 +767,104 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
 	return true
 end
 
+RegisterCustomEventListener("selection_update", function(data)
+	_G.selectionUnits[data.PlayerID] = data.entities
+end)
+
+
+RegisterCustomEventListener("courier_custom_select", function(data)
+	local pID = data.playerId
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
	local playerId = data.PlayerID
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 8]
```diff
@@ -687,6 +767,104 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
...
+		currentCourier = { _G.personalCouriers[pID]:GetEntityIndex() }
+	elseif _G.mainTeamCouriers[team]:IsAlive() then
+		currentCourier = { _G.mainTeamCouriers[team]:GetEntityIndex() }
+	end
+
+	if not currentCourier then return end
+	CustomGameEventManager:Send_ServerToPlayer(player, "selection_new", { entities = currentCourier })
+end)
+
+function unitMoveToPoint(unit, point)
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: Is there a reason not to use `CDOTA_BaseNPC:MoveToPosition(dest: Vector)`?

File: game/scripts/vscripts/addon_game_mode.lua [hunk 9]
```diff
@@ -483,15 +507,20 @@ function CMegaDotaGameMode:OnGameRulesStateChange(keys)
 
         end
 
-		local courier_spawn = {}
-		courier_spawn[2] = Entities:FindByClassname(nil, "info_courier_spawn_radiant")
-		courier_spawn[3] = Entities:FindByClassname(nil, "info_courier_spawn_dire")
+		if game_start then
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: Why is that required? State transition can happen only once

File: game/scripts/vscripts/addon_game_mode.lua [hunk 10]
```diff
@@ -687,6 +767,104 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
...
+
+function unitMoveToPoint(unit, point)
+	ExecuteOrderFromTable({
+		UnitIndex = unit:entindex(),
+		OrderType = DOTA_UNIT_ORDER_MOVE_TO_POSITION,
+		Position = point
+	})
+end
+
+RegisterCustomEventListener("courier_custom_select_deliever_items", function(data)
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: Wouldn't triggering `courier_transfer_items` or `courier_take_stash_and_transfer_items` work?

File: content/panorama/layout/custom_game/common/utils.js (hunk 4 above)
Comment (Sheodar) [reply content/panorama/layout/custom_game/common/utils.js]: Okay, I will put it in a separate file

File: game/scripts/vscripts/addon_game_mode.lua (hunk 8 above)
Comment (Sheodar) [reply game/scripts/vscripts/addon_game_mode.lua]: I missed this moment. I will replace it and check.

File: game/scripts/vscripts/addon_game_mode.lua (hunk 9 above)
Comment (Sheodar) [reply game/scripts/vscripts/addon_game_mode.lua]: I’m not sure how this works, but only in this case the skills work normally for couriers.

File: game/scripts/vscripts/addon_game_mode.lua (hunk 10 above)
Comment (Sheodar) [reply game/scripts/vscripts/addon_game_mode.lua]: I don’t understand, sry.

File: content/panorama/layout/custom_game/common/utils.js (hunk 5 above)
Comment (Sheodar) [reply content/panorama/layout/custom_game/common/utils.js]: Is it implicitly transmitted?

File: content/panorama/layout/custom_game/common/override_courier_ui.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+function FindDotaHudElement (id) {
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/override_courier_ui.js]: `FindDotaHudElement` and `GetDotaHud` are already defined in utils.js

File: content/panorama/layout/custom_game/common/utils.js [hunk 12]
```diff
@@ -66,3 +66,34 @@ function formatDate(date) {
...
+    var selectCourietButton = FindDotaHudElement('SelectCourierButton')
+    var deliverItemsButton = FindDotaHudElement('DeliverItemsButton')
+
+    selectCourietButton.SetPanelEvent("onactivate", function () {
+        GameEvents.SendCustomGameEventToServer("courier_custom_select", {})
+    })
+    deliverItemsButton.SetPanelEvent("onactivate", function () {
+        GameEvents.SendCustomGameEventToServer("courier_custom_select_deliever_items", {})
+    })
+})();
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/utils.js]: Should be removed

File: game/scripts/vscripts/addon_game_mode.lua [hunk 13]
```diff
@@ -776,8 +804,8 @@ RegisterCustomEventListener("GetKicks", function(data)
     CustomGameEventManager:Send_ServerToPlayer(PlayerResource:GetPlayer(data.id), "setkicks", {kicks = _G.kicks})
 end)
 
+
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]:
```suggestion
```

File: game/scripts/vscripts/addon_game_mode.lua [hunk 14]
```diff
@@ -795,6 +823,54 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
 	if disableHelpResult == false then
 		return false
 	end
+	CustomGameEventManager:Send_ServerToPlayer(PlayerResource:GetPlayer(playerId), "selection_refresh", { pID = playerId })
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: Is there a need to do that on any action?

File: game/scripts/vscripts/addon_game_mode.lua [hunk 15]
```diff
@@ -833,6 +909,69 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
...
+	local stashHasItems = false
+
+	for i = 9, 14 do
+		local item = player:GetAssignedHero():GetItemInSlot(i)
+		if item ~= nil then
+			stashHasItems = true
+		end
+	end
+
+	if stashHasItems then
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: Have you tried just always using one of these abilities?

File: game/scripts/vscripts/modifier_patreon_courier.lua [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+			MODIFIER_PROPERTY_FIXED_NIGHT_VISION,
+			MODIFIER_EVENT_ON_TAKEDAMAGE,
+			MODIFIER_PROPERTY_MODEL_CHANGE,
+			MODIFIER_EVENT_ON_MODEL_CHANGED,
+		}
+	end,
+	GetFixedDayVision = function() return 150 end,
+	GetFixedNightVision = function() return 150 end,
+
+	CheckState = function()
```
Comment (ark120202) [inline game/scripts/vscripts/modifier_patreon_courier.lua]: If it's not going to be used - remove it

File: game/scripts/vscripts/addon_game_mode.lua [hunk 17]
```diff
@@ -833,6 +909,69 @@ function CMegaDotaGameMode:ExecuteOrderFilter(filterTable)
 	return true
 end
 
+RegisterCustomEventListener("selection_update", function(data)
+	_G.selectionUnits[data.PlayerID] = data.entities
```
Comment (ark120202) [inline game/scripts/vscripts/addon_game_mode.lua]: What if client would send a huge list there?

File: content/panorama/layout/custom_game/common/override_courier_ui.js (hunk 11 above)
Comment (Sheodar) [reply content/panorama/layout/custom_game/common/override_courier_ui.js]: +

File: content/panorama/layout/custom_game/common/utils.js [hunk 18]
```diff
@@ -66,3 +66,21 @@ function formatDate(date) {
 		? date.getFullYear() + '-' + date.getMonth() + '-' + date.getDate()
 		: date.getMonth() + '/' + date.getDate() + '/' + date.getFullYear();
 }
+
+function FindDotaHudElement (id) {
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/utils.js]: It's already defined above

File: content/panorama/layout/custom_game/common/selection.js [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    // Skips units from the selected group
+    SelectionFilter(GetSelectedEntities())
+
+    //$.Schedule(0.03, SendSelectedEntities);
+}
+
+// Updates the list of selected entities on server for this player
+function SendSelectedEntities() {
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/selection.js]: There's some unused code now

File: content/panorama/layout/custom_game/common/override_courier_ui.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
+function SelectionCourierUpdate(msg) {
+    var needCourier = msg.newCourier;
+    var selected_entities = GetSelectedEntities();
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/override_courier_ui.js]:
```suggestion
    var selectedEntities = GetSelectedEntities();
```

File: content/panorama/layout/custom_game/common/override_courier_ui.js [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+    var haveCourierInSelect = false;
+    var selectionCounter = 0;
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/override_courier_ui.js]:
```suggestion
    var selectionCounter = selectedEntities.length;
```

File: content/panorama/layout/custom_game/common/override_courier_ui.js [hunk 22]
```diff
@@ -0,0 +1,35 @@
+function SelectionCourierUpdate(msg) {
+    var needCourier = msg.newCourier;
+    var selected_entities = GetSelectedEntities();
+    var haveCourierInSelect = false;
```
Comment (ark120202) [inline content/panorama/layout/custom_game/common/override_courier_ui.js]:
```suggestion
    var haveCourierInSelect = selectedEntities.some(Entities.IsCourier);
```

or

```suggestion
    var haveCourierInSelect = selectedEntities.some(function(e) { return Entities.IsCourier(e) });
```

would replace the loop

**Comment by darklordabc** [comment]
Need to add the item icon to the shop
